### DISCUSSION TITLE
Does composeRenderProps only work on client components?

### DISCUSSION THREAD

**Discussion by adneforslund** [adobe/react-spectrum#5767] [Q&A] [answered]
I have just started to use RAC with Next.js 14.1 and I have used the https://react-spectrum.adobe.com/react-aria-tailwind-starter as a starting point. But every time I use `composeRenderProps` in a component, I get the error 
`Unhandled Runtime Error Error: (0 , react_aria_components__WEBPACK_IMPORTED_MODULE_4__.composeRenderProps) is not a function
`, unless I specify the component as a client component. Is this the expected result?

**Comment by devongovett** [comment]
Probably? I guess since it returns a function that would need to be in a client component since functions can't be serialized.

**Reply by adneforslund** [reply]
@devongovett So if I read you correctly, RAC are generally meant to be client components?

**Reply by devongovett** [reply, marked as answer]
Yes, all React Aria components are client components because they use state and event handlers. I think this particular webpack error is potentially a next.js bug, but if you got rid of `composeRenderProps` and passed a function to `className`, you'd see a different one: "Functions cannot be passed directly to Client Components unless you explicitly expose it by marking it with "use server"." This means you need to make your component that's passing the function into a client component as well.

**Reply by abhion** [reply]
Hey, I can't find any documentation on what `composeRenderProps` does. Can you please tell me where to find it?

**Reply by ondrejkralik** [reply]
@abhion Imo is better to look into the source code
https://github.com/adobe/react-spectrum/blob/main/packages/react-aria-components/src/utils.tsx#L148

**Reply by oemer-aran** [reply]
I checked the code. But still struggle to understand `composeRenderProps`. Does it propagate it's own props to it's children? Can you explain it by example?
